core-bundle/src/Twig/Extension/ContaoExtension.php

```diff
@@ -0,0 +1,209 @@
+            // additionally support the Contao template hierarchy
+            new DynamicExtendsTokenParser($this->hierarchy),
+            new DynamicIncludeTokenParser($this->hierarchy),
+        ];
+    }
+
+    public function getFunctions(): array
+    {
+        $includeFunctionCallable = $this->getTwigIncludeFunction()->getCallable();
+
+        return [
+            // Overwrite the 'include' function to additionally support the
+            // Contao template hierarchy
+            new TwigFunction(
+                'include',
+                function (Environment $env, $context, $template, $variables = [], $withContext = true, $ignoreMissing = false, $sandboxed = false /* we need named arguments here */) use ($includeFunctionCallable) {
+                    $args = \func_get_args();
+                    $args[2] = DynamicIncludeTokenParser::adjustTemplateName((string) $template, $this->hierarchy);
+
+                    return $includeFunctionCallable(...$args);
+                },
+                ['needs_environment' => true, 'needs_context' => true, 'is_safe' => ['all']]
+            ),
+            new TwigFunction(
+                'contao_figure',
+                [FigureRendererRuntime::class, 'render'],
+                ['is_safe' => ['html']]
+            ),
+            new TwigFunction(
+                'picture_config',
+                [PictureConfigurationRuntime::class, 'fromArray']
+            ),
+            new TwigFunction(
+                'insert_tag',
+                [InsertTagRuntime::class, 'replace'],
+                ['is_safe' => ['html']]
+            ),
+            new TwigFunction(
+                'add_schema_org',
+                [SchemaOrgRuntime::class, 'add']
+            ),
+            new TwigFunction(
+                'contao_sections',
+                [LegacyTemplateFunctionsRuntime::class, 'renderLayoutSections'],
+                ['needs_context' => true, 'is_safe' => ['html']]
+            ),
+            new TwigFunction(
+                'contao_section',
+                [LegacyTemplateFunctionsRuntime::class, 'renderLayoutSection'],
+                ['needs_context' => true, 'is_safe' => ['html']]
+            ),
+            new TwigFunction(
+                'render_contao_backend_template',
+                [LegacyTemplateFunctionsRuntime::class, 'renderContaoBackendTemplate'],
+                ['is_safe' => ['html']]
+            ),
+        ];
+    }
+
+    /**
+     * @see \Contao\CoreBundle\Twig\Interop\PhpTemplateProxyNode
+     * @see \Contao\CoreBundle\Twig\Interop\PhpTemplateProxyNodeVisitor
+     *
+     * @internal
+     */
+    public function renderLegacyTemplate(string $name, array $blocks, array $context): string
+    {
+        $template = Path::getFilenameWithoutExtension($name);
+
+        $partialTemplate = new class($template) extends FrontendTemplate {
+            protected $blnEnableTwigSurrogateRendering = false;
```

Comment:
```suggestion
            protected function renderTwigSurrogateIfExists(): ?string
            {
                return null;
            }
```

Would this be a more clean approach? (Not sure though)

Reply:
I like this! :+1: 

Changed in 9edace8392c6dfd05c2a38b1b39361e9b74f7434.